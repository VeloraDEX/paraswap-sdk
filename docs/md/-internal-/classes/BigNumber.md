[**@paraswap/sdk**](../../README.md) • **Docs**

***

[@paraswap/sdk](../../globals.md) / [\<internal\>](../README.md) / BigNumber

# Class: BigNumber

## Implements

- [`Hexable`](../interfaces/Hexable.md)

## Constructors

### new BigNumber()

> **new BigNumber**(`constructorGuard`, `hex`): [`BigNumber`](BigNumber.md)

#### Parameters

• **constructorGuard**: `any`

• **hex**: `string`

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:7

## Properties

### \_hex

> `readonly` **\_hex**: `string`

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:5

***

### \_isBigNumber

> `readonly` **\_isBigNumber**: `boolean`

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:6

## Methods

### abs()

> **abs**(): [`BigNumber`](BigNumber.md)

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:10

***

### add()

> **add**(`other`): [`BigNumber`](BigNumber.md)

#### Parameters

• **other**: [`BigNumberish`](../type-aliases/BigNumberish.md)

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:11

***

### and()

> **and**(`other`): [`BigNumber`](BigNumber.md)

#### Parameters

• **other**: [`BigNumberish`](../type-aliases/BigNumberish.md)

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:17

***

### div()

> **div**(`other`): [`BigNumber`](BigNumber.md)

#### Parameters

• **other**: [`BigNumberish`](../type-aliases/BigNumberish.md)

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:13

***

### eq()

> **eq**(`other`): `boolean`

#### Parameters

• **other**: [`BigNumberish`](../type-aliases/BigNumberish.md)

#### Returns

`boolean`

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:23

***

### fromTwos()

> **fromTwos**(`value`): [`BigNumber`](BigNumber.md)

#### Parameters

• **value**: `number`

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:8

***

### gt()

> **gt**(`other`): `boolean`

#### Parameters

• **other**: [`BigNumberish`](../type-aliases/BigNumberish.md)

#### Returns

`boolean`

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:26

***

### gte()

> **gte**(`other`): `boolean`

#### Parameters

• **other**: [`BigNumberish`](../type-aliases/BigNumberish.md)

#### Returns

`boolean`

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:27

***

### isNegative()

> **isNegative**(): `boolean`

#### Returns

`boolean`

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:28

***

### isZero()

> **isZero**(): `boolean`

#### Returns

`boolean`

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:29

***

### lt()

> **lt**(`other`): `boolean`

#### Parameters

• **other**: [`BigNumberish`](../type-aliases/BigNumberish.md)

#### Returns

`boolean`

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:24

***

### lte()

> **lte**(`other`): `boolean`

#### Parameters

• **other**: [`BigNumberish`](../type-aliases/BigNumberish.md)

#### Returns

`boolean`

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:25

***

### mask()

> **mask**(`value`): [`BigNumber`](BigNumber.md)

#### Parameters

• **value**: `number`

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:20

***

### mod()

> **mod**(`other`): [`BigNumber`](BigNumber.md)

#### Parameters

• **other**: [`BigNumberish`](../type-aliases/BigNumberish.md)

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:15

***

### mul()

> **mul**(`other`): [`BigNumber`](BigNumber.md)

#### Parameters

• **other**: [`BigNumberish`](../type-aliases/BigNumberish.md)

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:14

***

### or()

> **or**(`other`): [`BigNumber`](BigNumber.md)

#### Parameters

• **other**: [`BigNumberish`](../type-aliases/BigNumberish.md)

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:18

***

### pow()

> **pow**(`other`): [`BigNumber`](BigNumber.md)

#### Parameters

• **other**: [`BigNumberish`](../type-aliases/BigNumberish.md)

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:16

***

### shl()

> **shl**(`value`): [`BigNumber`](BigNumber.md)

#### Parameters

• **value**: `number`

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:21

***

### shr()

> **shr**(`value`): [`BigNumber`](BigNumber.md)

#### Parameters

• **value**: `number`

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:22

***

### sub()

> **sub**(`other`): [`BigNumber`](BigNumber.md)

#### Parameters

• **other**: [`BigNumberish`](../type-aliases/BigNumberish.md)

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:12

***

### toBigInt()

> **toBigInt**(): `bigint`

#### Returns

`bigint`

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:31

***

### toHexString()

> **toHexString**(): `string`

#### Returns

`string`

#### Implementation of

[`Hexable`](../interfaces/Hexable.md).[`toHexString`](../interfaces/Hexable.md#tohexstring)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:33

***

### toJSON()

> **toJSON**(`key`?): `any`

#### Parameters

• **key?**: `string`

#### Returns

`any`

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:34

***

### toNumber()

> **toNumber**(): `number`

#### Returns

`number`

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:30

***

### toString()

> **toString**(): `string`

Returns a string representation of an object.

#### Returns

`string`

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:32

***

### toTwos()

> **toTwos**(`value`): [`BigNumber`](BigNumber.md)

#### Parameters

• **value**: `number`

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:9

***

### xor()

> **xor**(`other`): [`BigNumber`](BigNumber.md)

#### Parameters

• **other**: [`BigNumberish`](../type-aliases/BigNumberish.md)

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:19

***

### from()

> `static` **from**(`value`): [`BigNumber`](BigNumber.md)

#### Parameters

• **value**: `any`

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:35

***

### isBigNumber()

> `static` **isBigNumber**(`value`): `value is BigNumber`

#### Parameters

• **value**: `any`

#### Returns

`value is BigNumber`

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:36
