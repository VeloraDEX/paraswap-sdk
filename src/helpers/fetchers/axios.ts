import type { ExtraFetchParams, FetcherFunction } from '../../types';
import type { AxiosStatic } from 'axios';
import { FetcherError } from '../misc';

export type AxiosRequirement = Pick<AxiosStatic, 'request' | 'isAxiosError'>;

export const constructFetcher =
  (axios: AxiosRequirement, extra?: ExtraFetchParams): FetcherFunction =>
  async (params) => {
    try {
      const { requestParams, ...rest } = params;
      // adding apiKey to headers if it's provided
      const headers = extra?.apiKey
        ? {
            'X-API-KEY': extra.apiKey,
            ...rest.headers,
            ...requestParams?.headers,
          }
        : { ...rest.headers, ...requestParams };

      const allParams = { ...rest, ...requestParams, headers };

      const { data } = await axios.request(allParams);

      return data;
    } catch (error: any) {
      if (axios.isAxiosError(error)) {
        // attach a more meaningful message and other props
        throw new FetcherError(error);
      }

      throw error;
    }
  };
