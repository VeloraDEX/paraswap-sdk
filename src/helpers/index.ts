import { EthersV5ProviderDeps } from './providers/ethers';
import { EthersV6ProviderDeps } from './providers/ethersV6';
export { constructFetcher as constructAxiosFetcher } from './fetchers/axios';
export { constructFetcher as constructFetchFetcher } from './fetchers/fetch';
export {
  constructEthersV5ContractCaller as constructEthersContractCaller,
  constructEthersV5ContractCaller,
  EthersV5ProviderDeps,
} from './providers/ethers';
export {
  constructContractCaller as constructEthersV6ContractCaller,
  EthersV6ProviderDeps,
} from './providers/ethersV6';
export type EthersProviderDeps = EthersV5ProviderDeps | EthersV6ProviderDeps;

export {
  constructContractCaller as constructWeb3ContractCaller,
  Web3UnpromiEvent,
} from './providers/web3';
export {
  constructContractCaller as constructViemContractCaller,
  MinViemClient,
  txParamsToViemTxParams,
} from './providers/viem';
export { constructToken } from './token';
export { isFetcherError } from './misc';
export type { FetcherError, FetcherErrorInterface } from './misc';
