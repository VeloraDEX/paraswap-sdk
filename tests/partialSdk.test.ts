import * as dotenv from 'dotenv';
import Web3 from 'web3';
import { TransactionReceipt as Web3TransactionReceipt } from 'web3';
import { BigNumber as BigNumberEthers, ethers } from 'ethersV5';
import { ethers as ethersV6 } from 'ethers';
import axios from 'axios';
import fetch from 'isomorphic-unfetch';
import {
  ApproveTokenFunctions,
  BuildTxFunctions,
  constructApproveToken,
  constructAxiosFetcher,
  constructBuildTx,
  constructEthersV5ContractCaller,
  constructEthersV6ContractCaller,
  constructFetchFetcher,
  constructGetAdapters,
  constructGetRate,
  constructGetSpender,
  constructGetTokens,
  constructPartialSDK,
  constructWeb3ContractCaller,
  ConstructProviderFetchInput,
  GetSpenderFunctions,
  GetTokensFunctions,
  SDKConfig,
  Web3UnpromiEvent,
  constructGetBalances,
  GetBalancesFunctions,
  GetAdaptersFunctions,
  GetRateFunctions,
  isAllowance,
  SwapSide,
  GetSwapTxFunctions,
  constructSwapTx,
} from '../src';
import BigNumber from 'bignumber.js';

import erc20abi from './abi/ERC20.json';

import { assert } from 'ts-essentials';
import type {
  ContractCallerFunctions,
  StaticContractCallerFn,
  TransactionContractCallerFn,
} from '../src/types';
import { HardhatProvider, setupFork } from './helpers/hardhat';

dotenv.config();

jest.setTimeout(30 * 1000);

const ETH = '0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee';
const DAI = '0x6B175474E89094C44Da98b954EedeAC495271d0F';
// const HEX = '0x2b591e99afe9f32eaa6214f7b7629768c40eeb39';

const DUMMY_ADDRESS_FOR_TESTING_ALLOWANCES =
  '0xb9A079479A7b0F4E7F398F7ED3946bE6d9a40E79';

const chainId = 1;
const srcToken = ETH;
const destToken = DAI;
const srcAmount = (1 * 1e18).toString(); //The source amount multiplied by its decimals

const referrer = 'sdk-test';

const TEST_MNEMONIC =
  'radar blur cabbage chef fix engine embark joy scheme fiction master release';
//0xaC39b311DCEb2A4b2f5d8461c1cdaF756F4F7Ae9
const wallet = ethers.Wallet.fromMnemonic(TEST_MNEMONIC);
const walletV6 = ethersV6.HDNodeWallet.fromPhrase(TEST_MNEMONIC);

const web3provider = new Web3(HardhatProvider as any);

const ethersProvider = new ethers.providers.Web3Provider(
  HardhatProvider as any
);

const ethersV6Provider = new ethersV6.BrowserProvider(HardhatProvider);
const signerV6 = walletV6.connect(ethersV6Provider);

const fetchFetcher = constructFetchFetcher(fetch);
const axiosFetcher = constructAxiosFetcher(axios);

const signer = wallet.connect(ethersProvider);
const senderAddress = signer.address;

const ethersV5ContractCaller = constructEthersV5ContractCaller(
  {
    ethersProviderOrSigner: signer,
    EthersContract: ethers.Contract,
  },
  senderAddress
);

const ethersV6ContractCaller = constructEthersV6ContractCaller(
  {
    ethersV6ProviderOrSigner: signerV6,
    EthersV6Contract: ethersV6.Contract,
  },
  senderAddress
);

const web3ContractCaller = constructWeb3ContractCaller(
  web3provider,
  senderAddress
);

const customGanacheContractCaller = constructProviderOnlyContractCaller(
  HardhatProvider,
  senderAddress
);

describe.each([
  ['fetchFetcher', fetchFetcher],
  ['axiosFetcher', axiosFetcher],
])('ParaSwap SDK: fetching methods: %s', (testName, fetcher) => {
  let paraSwap: GetBalancesFunctions &
    GetAdaptersFunctions &
    GetTokensFunctions &
    GetRateFunctions &
    GetSpenderFunctions &
    BuildTxFunctions &
    GetSwapTxFunctions;

  beforeAll(async () => {
    await setupFork({
      accounts: [{ address: senderAddress, balance: 8e18 }],
    });

    paraSwap = constructPartialSDK(
      { chainId, fetcher, version: '5' },
      constructGetBalances,
      constructGetAdapters,
      constructGetTokens,
      constructGetRate,
      constructGetSpender,
      constructBuildTx,
      constructSwapTx
    );
  });
  test('getBalance', async () => {
    try {
      const balance = await paraSwap.getBalance(senderAddress, ETH);
      expect(balance).toBeDefined();
    } catch (error: any) {
      // workaround for API sometimes failing on some Tokens(?)
      expect(error.message).toMatch(/Only chainId \d+ is supported/);
    }
  });

  test('Get_Markets', async () => {
    const markets = await paraSwap.getAdapters();
    expect(markets.length).toBeGreaterThan(15);
  });

  test('Get_Tokens', async () => {
    const tokens = await paraSwap.getTokens();

    expect(Array.isArray(tokens)).toBe(true);
    expect(tokens.length).toBeGreaterThan(0);
    expect(tokens[0]).toEqual(
      expect.objectContaining({
        symbol: expect.any(String),
        address: expect.any(String),
        decimals: expect.any(Number),
      })
    );
  });

  test('Get_Rates', async () => {
    const priceRoute = await paraSwap.getRate({
      srcToken: ETH,
      destToken: DAI,
      amount: srcAmount,
      userAddress: senderAddress,
      side: SwapSide.SELL,
      options: {
        includeDEXS: ['UniswapV2'],
        otherExchangePrices: true,
      },
    });

    const { destAmount, bestRoute, others } = priceRoute;

    expect(typeof destAmount).toBe('string');

    expect(Array.isArray(bestRoute)).toBe(true);

    const swapExchange = bestRoute[0]?.swaps[0]?.swapExchanges[0];

    assert(swapExchange, 'exchange available at swapExchanges[0]');

    expect(typeof swapExchange.destAmount).toBe('string');
    expect(new BigNumber(swapExchange.destAmount).isNaN()).toBe(false);

    expect(typeof swapExchange.exchange).toBe('string');

    const firstBestRoute = bestRoute[0];
    assert(firstBestRoute, 'route available at bestRoute[0]');

    expect(typeof firstBestRoute.percent).toBe('number');
    expect(new BigNumber(firstBestRoute.percent).isNaN()).toBe(false);

    expect(typeof swapExchange.srcAmount).toBe('string');
    expect(new BigNumber(swapExchange.srcAmount).isNaN()).toBe(false);

    expect(Array.isArray(others)).toBe(true);

    const firstRoute = others?.[0];

    assert(firstRoute, 'at least one route must exist');

    expect(typeof firstRoute.exchange).toBe('string');

    expect(typeof firstRoute.unit).toBe('string');
    expect(firstRoute.unit && new BigNumber(firstRoute.unit).isNaN()).toBe(
      false
    );
  });

  test('Get_SwapTxData', async () => {
    const { priceRoute, txParams } = await paraSwap.getSwapTxData({
      srcToken: ETH,
      destToken: DAI,
      amount: srcAmount,
      userAddress: senderAddress,
      side: SwapSide.SELL,
      slippage: 500,
      options: {
        includeDEXS: ['UniswapV2'],
      },
    });

    const bestRouteStable = priceRoute.bestRoute.map((b) => ({
      ...b,
      swaps: b.swaps.map((s) => ({
        ...s,
        swapExchanges: s.swapExchanges.map((se) => ({
          ...se,
          destAmount: 'dynamic_number',
          data: {
            ...se.data,
            gasUSD: 'dynamic_number',
          },
        })),
      })),
    }));

    const priceRouteStable = {
      ...priceRoute,
      gasCost: 'dynamic_number',
      gasCostUSD: 'dynamic_number',
      hmac: 'dynamic_number',
      destAmount: 'dynamic_number',
      blockNumber: 'dynamic_number',
      srcUSD: 'dynamic_number',
      destUSD: 'dynamic_number',
      bestRoute: bestRouteStable,
    };

    const txParamsStable = {
      ...txParams,
      data: 'dynamic_string',
      from: 'dynamic_string',
      gasPrice: 'dynamic_number',
    };

    expect(txParams.from).toEqual(senderAddress);

    expect(priceRouteStable).toMatchSnapshot('Get_SwapTxData::priceRoute');
    expect(txParamsStable).toMatchSnapshot('Get_SwapTxData::txParams');
  });

  test('Get_Spender', async () => {
    const spender = await paraSwap.getSpender();
    expect(web3provider.utils.isAddress(spender)).toBe(true);
  });

  test('Get_AugustusSwapper', async () => {
    const Augustus = await paraSwap.getAugustusSwapper();
    expect(web3provider.utils.isAddress(Augustus)).toBe(true);
  });

  test('Get_Contracts', async () => {
    const contracts = await paraSwap.getContracts();
    expect(contracts).toMatchSnapshot(testName);
  });

  test('Get_Allowance', async () => {
    const allowance = await paraSwap.getAllowance(
      DUMMY_ADDRESS_FOR_TESTING_ALLOWANCES,
      DAI
    );

    assert(isAllowance(allowance), 'hardcoded dummy address should be found');

    expect(allowance.allowance).toEqual('123000000000000000');
  });

  test('Get_Adapters', async () => {
    const adapters = await paraSwap.getAdapters();

    expect(adapters).toMatchSnapshot('Get_Adapters');
  });

  test('Build_Tx', async () => {
    const destToken = '0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48';
    const priceRoute = await paraSwap.getRate({
      srcToken,
      destToken,
      amount: srcAmount,
      userAddress: senderAddress,
      side: SwapSide.SELL,
      options: {
        includeDEXS: ['UniswapV2'],
      },
    });

    const destAmount = new BigNumber(priceRoute.destAmount)
      .times(0.99)
      .toFixed(0);

    const txParams = await paraSwap.buildTx(
      {
        srcToken,
        destToken,
        srcAmount,
        destAmount,
        priceRoute,
        userAddress: senderAddress,
        partner: referrer,
      },
      { ignoreChecks: true }
    );

    expect(typeof txParams).toBe('object');
  });
  test('Build_and_Send_Tx', async () => {
    const priceRoute = await paraSwap.getRate({
      srcToken,
      destToken,
      amount: srcAmount,
      userAddress: senderAddress,
      side: SwapSide.SELL,
      options: {
        includeDEXS: ['Uniswap', 'UniswapV2', 'Balancer', 'Oasis'],
      },
    });

    const destAmount = new BigNumber(priceRoute.destAmount)
      .times(0.99)
      .toFixed(0);

    const txParams = await paraSwap.buildTx(
      {
        srcToken,
        destToken,
        srcAmount,
        destAmount,
        priceRoute,
        userAddress: signer.address,
        partner: referrer,
      },
      { ignoreChecks: true }
    );

    const transaction = {
      ...txParams,
      gasPrice:
        txParams.gasPrice &&
        '0x' + new BigNumber(txParams.gasPrice).toString(16),
      maxFeePerGas:
        txParams.maxFeePerGas &&
        '0x' + new BigNumber(txParams.maxFeePerGas).toString(16),
      maxPriorityFeePerGas:
        txParams.maxPriorityFeePerGas &&
        '0x' + new BigNumber(txParams.maxPriorityFeePerGas).toString(16),
      gasLimit: '0x' + new BigNumber(5000000).toString(16),
      value: '0x' + new BigNumber(txParams.value).toString(16),
    };
    const toContract = new ethers.Contract(destToken, erc20abi, ethersProvider);
    const beforeFromBalance = await ethersProvider.getBalance(signer.address);
    const beforeToBalance = await toContract.balanceOf(signer.address);

    const txr = await signer.sendTransaction(transaction);
    await txr.wait(1);
    const afterFromBalance = await ethersProvider.getBalance(signer.address);
    const afterToBalance = await toContract.balanceOf(signer.address);
    expect(beforeFromBalance.gt(afterFromBalance)).toBeTruthy();
    expect(beforeToBalance.lt(afterToBalance)).toBeTruthy();
  });
  test('Build_and_Send_Tx_BUY', async () => {
    const destAmount = srcAmount;
    const priceRoute = await paraSwap.getRate({
      srcToken,
      destToken,
      amount: destAmount,
      userAddress: senderAddress,
      side: SwapSide.BUY,
      options: { includeDEXS: ['Uniswap', 'UniswapV2', 'Balancer', 'Oasis'] },
    });
    const _srcAmount = new BigNumber(priceRoute.srcAmount)
      .times(1.1)
      .toFixed(0);

    const txParams = await paraSwap.buildTx(
      {
        srcToken,
        destToken,
        srcAmount: _srcAmount,
        destAmount,
        priceRoute,
        userAddress: signer.address,
        partner: referrer,
      },
      { ignoreChecks: true }
    );

    const transaction = {
      ...txParams,
      gasPrice:
        txParams.gasPrice &&
        '0x' + new BigNumber(txParams.gasPrice).toString(16),
      maxFeePerGas:
        txParams.maxFeePerGas &&
        '0x' + new BigNumber(txParams.maxFeePerGas).toString(16),
      maxPriorityFeePerGas:
        txParams.maxPriorityFeePerGas &&
        '0x' + new BigNumber(txParams.maxPriorityFeePerGas).toString(16),
      gasLimit: '0x' + new BigNumber(5000000).toString(16),
      value: '0x' + new BigNumber(txParams.value).toString(16),
    };
    const toContract = new ethers.Contract(destToken, erc20abi, ethersProvider);
    const beforeFromBalance = await ethersProvider.getBalance(signer.address);
    const beforeToBalance = await toContract.balanceOf(signer.address);

    const txr = await signer.sendTransaction(transaction);
    await txr.wait(1);
    const afterFromBalance = await ethersProvider.getBalance(signer.address);
    const afterToBalance = await toContract.balanceOf(signer.address);
    expect(beforeFromBalance.gt(afterFromBalance)).toBeTruthy();
    expect(beforeToBalance.lt(afterToBalance)).toBeTruthy();
  });
});

describe.each([
  [
    'fetchFetcher & ethersV5ContractCaller',
    fetchFetcher,
    ethersV5ContractCaller,
  ],
  [
    'fetchFetcher & ethersV6ContractCaller',
    fetchFetcher,
    ethersV6ContractCaller,
  ],
  ['axiosFetcher & web3ContractCaller', axiosFetcher, web3ContractCaller],
  [
    'axiosFetcher & customGanacheContractCaller',
    axiosFetcher,
    customGanacheContractCaller,
  ],
])(
  'ParaSwap SDK: contract calling methods: %s',
  (testName, fetcher, contractCaller) => {
    type ApproveTxResult =
      | ethers.ContractTransaction
      | ethersV6.ContractTransactionResponse
      | Web3UnpromiEvent
      | string;
    // @TODO try Instantiation Expression when TS 4.7 `as constructApproveToken<TxResponse>`
    type ApproveConstructor = (
      options: ConstructProviderFetchInput<ApproveTxResult, 'transactCall'>
    ) => ApproveTokenFunctions<ApproveTxResult>;

    let paraSwap: ApproveTokenFunctions<ApproveTxResult> & GetSpenderFunctions;

    beforeAll(() => {
      paraSwap = constructPartialSDK<
        SDKConfig<ApproveTxResult>,
        [ApproveConstructor, typeof constructGetSpender]
      >(
        { chainId, fetcher, contractCaller, version: '5' },
        constructApproveToken,
        constructGetSpender
      );
    });
    test('approveToken', async () => {
      const tx = await paraSwap.approveToken('12345', DAI);

      if (typeof tx === 'string') {
        // tx is sent
        // wait for it to mine however you prefer
        await ethersProvider.waitForTransaction(tx);
      } else if ('wait' in tx) {
        await tx.wait(1);
      } else if ('on' in tx) {
        await new Promise<Web3TransactionReceipt>((resolve, reject) => {
          tx.once('receipt', resolve);
          tx.once('error', reject);
        });
      }
      const toContract = new ethers.Contract(
        destToken,
        erc20abi,
        ethersProvider
      );
      const spender = await paraSwap.getSpender();
      const allowance: BigNumberEthers = await toContract.allowance(
        signer.address,
        spender
      );
      expect(allowance.toString()).toEqual('12345');
    });
  }
);

interface MinProvider {
  request(args: { method: string; params: any }): Promise<any>;
}
// example of constructing a custom contractCaller with provider only
function constructProviderOnlyContractCaller(
  provider: MinProvider,
  account?: string
): Pick<
  ContractCallerFunctions<string>,
  'staticCall' | 'transactCall' | 'signTypedDataCall'
> {
  // staticCall isn't currently necessary, because provider is only used in approveToken currently for tx making
  const staticCall: StaticContractCallerFn = async ({
    address,
    abi,
    contractMethod,
    args,
    overrides,
  }) => {
    // need to encode data somehow, to send raw call
    const iface = new ethers.utils.Interface(abi);
    const calldataEncoded = iface.encodeFunctionData(contractMethod, args);

    const gasPrice =
      overrides.gasPrice &&
      '0x' + Number.parseInt(overrides.gasPrice).toString(16);
    const gas = overrides.gas && '0x' + overrides.gas.toString(16);
    const value = overrides.value && '0x' + overrides.value.toString(16);

    const params = [
      {
        from: account,
        to: address,
        gasPrice,
        gas,
        value,
        data: calldataEncoded,
      },
    ];

    // what provider here returns will depend on provider
    // ganache returns `result` from {"id":1, "jsonrpc": "2.0", "result": "0x..."} JsonRpcResponse
    // keeping it `as any` as we don't know
    const res = await provider.request({
      method: 'eth_call',
      params,
    });

    return res;
  };

  const transactCall: TransactionContractCallerFn<string> = async ({
    address,
    abi,
    contractMethod,
    args,
    overrides,
  }) => {
    const iface = new ethers.utils.Interface(abi);
    const calldataEncoded = iface.encodeFunctionData(contractMethod, args);

    const gasPrice =
      overrides.gasPrice &&
      '0x' + Number.parseInt(overrides.gasPrice).toString(16);
    const gas = overrides.gas && '0x' + overrides.gas.toString(16);
    const value = overrides.value && '0x' + overrides.value.toString(16);

    const params = [
      {
        from: account,
        to: address,
        gasPrice,
        gas,
        value,
        data: calldataEncoded,
      },
    ];

    // what provider here returns will depend on provider
    // hardhat returns `result` from {"id":1, "jsonrpc": "2.0", "result": "0x..."} JsonRpcResponse
    // `string`
    const res = await provider.request({
      // we can only sendTransaction right away because we asked hardhat to impersonate account
      // for other provider we may need to signTransaction first, then sendRawTransaction
      method: 'eth_sendTransaction',
      params,
    });

    return res;
  };

  const signTypedDataCall = () => {
    throw new Error('not implemented');
  };

  return { transactCall, signTypedDataCall, staticCall };
}
