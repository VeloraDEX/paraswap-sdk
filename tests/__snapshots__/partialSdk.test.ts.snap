// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ParaSwap SDK: fetching methods: axiosFetcher Get_Adapters: Get_Adapters 1`] = `
Array [
  "Bancor",
  "Compound",
  "Fulcrum",
  "AugustusRFQ",
  "BDai",
  "idle",
  "Shell",
  "DODOV1",
  "DODOV2",
  "TraderJoeV2.1",
  "Smoothy",
  "OneInchLP",
  "CurveV2",
  "mStable",
  "BProtocol",
  "Jarvis",
  "Lido",
  "EtherFi",
  "CurveV1",
  "Swerve",
  "BalancerV1",
  "BalancerV2",
  "UniswapV2",
  "DefiSwap",
  "LinkSwap",
  "PancakeSwapV2",
  "SushiSwap",
  "Swapsicle",
  "ShibaSwap",
  "SakeSwap",
  "Verse",
  "UniswapV3",
  "SushiSwapV3",
  "PancakeswapV3",
  "AaveV1",
  "AaveV2",
  "AaveV3",
  "KyberDmm",
  "Weth",
  "PolygonMigrator",
  "MakerPsm",
  "Saddle",
  "Synapse",
  "ParaSwapLimitOrders",
  "SolidlyV2",
  "Synthetix",
  "CurveV1Factory",
  "wstETH",
  "Hashflow",
  "MaverickV1",
  "SwaapV2",
  "AngleStakedStableUSD",
  "AngleStakedStableEUR",
  "SolidlyV3",
  "Smardex",
  "Wombat",
  "Swell",
  "Spark",
]
`;

exports[`ParaSwap SDK: fetching methods: axiosFetcher Get_Contracts: axiosFetcher 1`] = `
Object {
  "AugustusRFQ": "0xe92b586627ccA7a83dC919cc7127196d70f55a06",
  "AugustusSwapper": "0xDEF171Fe48CF0115B1d80b88dc8eAB59176FEe57",
  "TokenTransferProxy": "0x216b4b4ba9f3e719726886d34a177484278bfcae",
}
`;

exports[`ParaSwap SDK: fetching methods: fetchFetcher Get_Adapters: Get_Adapters 1`] = `
Array [
  "Bancor",
  "Compound",
  "Fulcrum",
  "AugustusRFQ",
  "BDai",
  "idle",
  "Shell",
  "DODOV1",
  "DODOV2",
  "TraderJoeV2.1",
  "Smoothy",
  "OneInchLP",
  "CurveV2",
  "mStable",
  "BProtocol",
  "Jarvis",
  "Lido",
  "EtherFi",
  "CurveV1",
  "Swerve",
  "BalancerV1",
  "BalancerV2",
  "UniswapV2",
  "DefiSwap",
  "LinkSwap",
  "PancakeSwapV2",
  "SushiSwap",
  "Swapsicle",
  "ShibaSwap",
  "SakeSwap",
  "Verse",
  "UniswapV3",
  "SushiSwapV3",
  "PancakeswapV3",
  "AaveV1",
  "AaveV2",
  "AaveV3",
  "KyberDmm",
  "Weth",
  "PolygonMigrator",
  "MakerPsm",
  "Saddle",
  "Synapse",
  "ParaSwapLimitOrders",
  "SolidlyV2",
  "Synthetix",
  "CurveV1Factory",
  "wstETH",
  "Hashflow",
  "MaverickV1",
  "SwaapV2",
  "AngleStakedStableUSD",
  "AngleStakedStableEUR",
  "SolidlyV3",
  "Smardex",
  "Wombat",
  "Swell",
  "Spark",
]
`;

exports[`ParaSwap SDK: fetching methods: fetchFetcher Get_Contracts: fetchFetcher 1`] = `
Object {
  "AugustusRFQ": "0xe92b586627ccA7a83dC919cc7127196d70f55a06",
  "AugustusSwapper": "0xDEF171Fe48CF0115B1d80b88dc8eAB59176FEe57",
  "TokenTransferProxy": "0x216b4b4ba9f3e719726886d34a177484278bfcae",
}
`;
